src/bcf/report/report.rs

```diff
@@ -0,0 +1,527 @@
+    v.sort_by(|(a, b), (c, d)| if b == d { a.cmp(&c) } else { d.cmp(&b) });
+    let ordered_genes: Vec<_> = v.iter().map(|(x, _)| x).collect();
+    let order = json!(ordered_genes);
+
+    let pages = v.len() / page_size;
+
+    for i in 0..pages + 1 {
+        let current_genes = if i != pages {
+            &v[(i * page_size)..((i + 1) * page_size)] // get genes for current page
+        } else {
+            &v[(i * page_size)..] // get genes for last page
+        };
+
+        let mut sorted_genes = Vec::new();
+        for (g, _) in current_genes {
+            sorted_genes.push(g);
+        }
+
+        let page = i + 1;
+
+        let page_data: Vec<_> = data
+            .iter()
+            .filter(|entry| sorted_genes.contains(&&entry.gene))
+            .sorted()
+            .collect();
+
+        let impact_page_data: Vec<_> = impact_data
+            .iter()
+            .filter(|entry| sorted_genes.contains(&&entry.record.key))
+            .sorted()
+            .collect();
+
+        let consequence_page_data: Vec<_> = consequence_data
+            .iter()
+            .filter(|entry| sorted_genes.contains(&&entry.record.key))
+            .sorted()
+            .collect();
+
+        let clin_sig_page_data: Vec<_> = clin_sig_data
+            .iter()
+            .filter(|entry| sorted_genes.contains(&&entry.record.key))
+            .sorted()
+            .collect();
+
+        let af_page_data: Vec<_> = af_data
+            .iter()
+            .filter(|entry| sorted_genes.contains(&&entry.key))
+            .collect();
+
+        let mut vl_specs: Value = serde_json::from_str(include_str!("report_specs.json")).unwrap();
+        let values = json!({"main": page_data , "impact": impact_page_data, "consequence": consequence_page_data , "clin_sig": clin_sig_page_data, "allel_frequency": af_page_data});
+
+        vl_specs["datasets"] = values;
+
+        let mut packer = Packer::new();
+        let options = PackOptions::new();
+        let packed_specs = packer.pack(&vl_specs, &options).unwrap();
+        let mut templates = Tera::default();
+        templates.register_filter("embed_source", embed_source);
+        templates.add_raw_template("report.html.tera", include_str!("report.html.tera"))?;
+        let mut context = Context::new();
+        let data = serde_json::to_string(&packed_specs)?;
+        context.insert("oncoprint", &data);
+        context.insert("pages", &(pages + 1));
+        context.insert("order", &serde_json::to_string(&order)?);
+        let local: DateTime<Local> = Local::now();
+        context.insert("time", &local.format("%a %b %e %T %Y").to_string());
+        context.insert("version", &env!("CARGO_PKG_VERSION"));
+
+        let html = templates.render("report.html.tera", &context)?;
+
+        let index = output_path.to_owned() + "/index" + &page.to_string() + ".html";
+        let mut file = File::create(index)?;
+        file.write_all(html.as_bytes())?;
+    }
+
+    Ok(())
```

Comment:
This entire function is very long. Please split it into smaller pieces, such that it becomes more readable.

Reply:
I moved the rendering of the second stage into a separate function. Otherwise it might be a bit difficult since most of the code is juggling around with the data. Moving these parts also into seperate functions would end up with big return tuples.